src/rules/max-expects.ts

```diff
@@ -0,0 +1,89 @@
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import {
+  FunctionExpression,
+  createRule,
+  isExpectCall,
+  isTypeOfJestFnCall,
+} from './utils';
+
+export default createRule({
+  name: __filename,
+  meta: {
+    docs: {
+      category: 'Best Practices',
+      description: 'Enforces a maximum assertion calls in a test body',
+      recommended: false,
+    },
+    messages: {
+      exceededMaxAssertion:
+        'Too many assertion calls ({{ count }}). Maximum allowed is {{ max }}.',
+    },
+    type: 'suggestion',
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          max: {
+            type: 'integer',
+            minimum: 0,
+          },
+        },
+        additionalProperties: false,
+      },
+    ],
+  },
+  defaultOptions: [{ max: 5 }],
+  create(context, [{ max }]) {
+    let count = 0;
+
+    const onFunctionExpressionEnter = (node: FunctionExpression) => {
+      if (!node?.parent) {
+        return;
+      }
+
+      const isTestFn =
+        node.parent.type !== AST_NODE_TYPES.CallExpression ||
+        isTypeOfJestFnCall(node.parent, context, ['test']);
+
+      if (isTestFn) {
+        count = 0;
+
+        return;
+      }
+    };
+    const onFunctionExpressionExit = (node: FunctionExpression) => {
+      if (!node?.parent) {
+        return;
+      }
+
+      const isTestFn =
+        node.parent.type !== AST_NODE_TYPES.CallExpression ||
+        isTypeOfJestFnCall(node.parent, context, ['test']);
+
+      if (isTestFn) {
+        count = count - 1;
+
+        return;
+      }
+    };
+
+    return {
+      FunctionExpression: onFunctionExpressionEnter,
+      'FunctionExpression:exit': onFunctionExpressionExit,
+      ArrowFunctionExpression: onFunctionExpressionEnter,
+      'ArrowFunctionExpression:exit': onFunctionExpressionExit,
+      CallExpression(node) {
+        if (isExpectCall(node)) {
+          count += 1;
+        }
+        if (count > max && node.parent) {
+          context.report({
+            node: node.parent,
```

Comment:
nit: I'm not sure if we should be reporting on the parent or the node itself - it doesn't matter too much except it lets us drop an impossible condition (TS doesn't know that `parent` is always present except for `Program` type nodes).

Reply:
Fixed in https://github.com/jest-community/eslint-plugin-jest/pull/1166/commits/29174fd4aae37610c9b869ec836d336cb7bd174f and https://github.com/jest-community/eslint-plugin-jest/pull/1166/commits/589b7d48f0abe0f70fe8d37d01f8c3eee5d17d0a.